File: intellij-lsp/src/com/github/gtache/actions/LSPReformatAction.scala
package com.github.gtache.actions

import java.util.concurrent.CompletableFuture
import java.util.function.Supplier

import com.github.gtache.PluginMain
import com.github.gtache.requests.ReformatHandler
import com.intellij.codeInsight.actions.ReformatCodeAction
import com.intellij.lang.LanguageFormatting
import com.intellij.openapi.actionSystem.{AnActionEvent, CommonDataKeys}
import com.intellij.openapi.diagnostic.Logger
import com.intellij.psi.PsiDocumentManager

class LSPReformatAction extends ReformatCodeAction {

  private val LOG: Logger = Logger.getInstance(classOf[LSPReformatAction])

  override def actionPerformed(e: AnActionEvent): Unit = {
    val project = e.getData(CommonDataKeys.PROJECT)
    val editor = e.getData(CommonDataKeys.EDITOR)
    if (editor != null) {
      val file = PsiDocumentManager.getInstance(project).getPsiFile(editor.getDocument)
      if (!LanguageFormatting.INSTANCE.allForLanguage(file.getLanguage).isEmpty) {
        super.actionPerformed(e)
      } else if (PluginMain.isExtensionSupported(file.getVirtualFile.getExtension)) {
        ReformatHandler.reformatFile(editor)
      }
    } else if (project != null) {
      CompletableFuture.supplyAsync(new Supplier[Boolean] {
        override def get(): Boolean = ReformatHandler.reformatAllFiles(project)
      }).thenAccept(res => if (!res) super.actionPerformed(e))
    } else {
      super.actionPerformed(e)
    }
  }

  override def update(event: AnActionEvent): Unit = super.update(event)

}

File: intellij-lsp/src/main/kotlin/com/github/gtache/lsp/actions/LSPReformatAction.kt
package com.github.gtache.lsp.actions

import com.github.gtache.lsp.LSPProjectService
import com.github.gtache.lsp.requests.ReformatHandler
import com.github.gtache.lsp.settings.LSPProjectSettings
import com.intellij.codeInsight.actions.ReformatCodeAction
import com.intellij.lang.LanguageFormatting
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.CommonDataKeys
import com.intellij.openapi.components.service
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.project.DumbAware
import com.intellij.psi.PsiDocumentManager

/**
 * Action overriding the default reformat action
 * Fallback to the default action if the language is already supported or not supported by any language server
 */
class LSPReformatAction : ReformatCodeAction(), DumbAware {

    companion object {
        private val logger: Logger = Logger.getInstance(LSPReformatAction::class.java)
    }

    override fun actionPerformed(e: AnActionEvent): Unit {
        val project = e.getData(CommonDataKeys.PROJECT)
        val editor = e.getData(CommonDataKeys.EDITOR)
        if (editor != null && project != null) {
            val file = PsiDocumentManager.getInstance(project).getPsiFile(editor.document)
            val state = project.service<LSPProjectSettings>().projectState
            if (file != null && (state.isAlwaysSendRequests ||
                        (LanguageFormatting.INSTANCE.allForLanguage(file.language).isEmpty()
                                && project.service<LSPProjectService>().isExtensionSupported(file.virtualFile.extension)))
            ) {
                ReformatHandler.reformatFile(editor)
            } else {
                super.actionPerformed(e)
            }
        }
    }

}